### ISSUE TITLE
modX::makeUrl and modX::getContext cache a wrong key/context relation

### ISSUE THREAD

**Issue by CyntaxDotDev** [modxbot/migrate#9769]
ReSpawN created Redmine issue ID 9769

The following bit of code is responsible for fetching the context linked to the given ID. Then, if found (which in this case works fine), it fetches the context via modX::getContext which in turn returns the context (prepared) object.

modx.class.php, function modX::makeURL at line ~#926

<pre>
if ($context == '') {
    if ($results = $this->query("SELECT context_key FROM " . $this->getTableName('modResource') . " WHERE id = {$id}")) {
        $contexts= $results->fetchAll(PDO::FETCH_COLUMN);
        if ($contextKey = reset($contexts)) {
            $ctx = $this->getContext($contextKey);
        }
    }
} else {
    $ctx = $this->getContext($context);
}
</pre>

This is where things get fishy. When I dump the array with print_r(array_keys($modx->contexts)) inside the function, I get only one key, web, which is fine. Now, when I dump that key, inside of it, is a whole different context! Namely, the context I'm currently in. For the sake of argument, let's call it "demo" (and of course "web").

What I am trying to accomplish is simply generate a LINK inside a resource placed inside "demo" to a resource in "web". I've got all the rights, both logged in and logged out don't work, fowarding across contexts enabled and still it doesn't work.

While that might be an oversight on my part, which I still haven't found, getContext fetching a context with key "web" which contains "demo" as an object is clearly wrong.

I'm switching contexts based on domain, in an extension package which switches while still in the construct via a custom _initContext.

**Comment by CyntaxDotDev** [comment]
ReSpawN submitted:

Got a small update. I am using a Wayfinder call in front of the makeUrl ([[~42]]) portion, and when I remove startIdContext and/or contexts as arguments, the links are generated fine...

**Comment by pyrographics** [comment]
pyrographics submitted:

I have a similar issue on 2.2.7.  site1.com with site_start of resource 1 and site2.com with site_start of resource 2 will not cross link to each other. [[~1]] on site2.com incorrectly links back to site2.com when it should link to site1.com, [[~2]] on site1.com incorrectly links back to site1.com. Both sites have the base_url and site_url set in their respective context settings. Why are they not getting the site_url from the context that contains the resource? I have allow_forward_across_contexts turned on so that I can use symlinks to share the same content on some pages if that matters.  I have tested it on two different manager installations, one using the ContextRouter package and the other using the Gateway Manager package.  I did not mess with Access Policies on either.
